CORDA-2942: Node lifecycle events

vkolomeyko opened the pull request:
Lifecycle events for `CordaService`s, please see linked Jira for the motivation for this change.

Suggest to start review from documentation changes: `api-service-classes.rst` then move on to the integration test `CordaServiceLifecycleTests` and then have a look at `ServiceLifecycleObserver`.

Definitely needs squash/rebase when integrating this PR.

Commit:
CORDA-2942: Port minimal set of changes to make lifecycle events work

... and make codebase compile.

Commit:
CORDA-2942: Undo some changes which are not strictly speaking necessary

Commit:
CORDA-2942: Make `NodeServicesContext` leaner and delete `extensions-api` module

Commit:
CORDA-2942: Reduce even more number of files affected

Commit:
CORDA-2942: Integration test fix

Commit:
CORDA-2942: Make events `AfterStart` and `BeforeStop` generic w.r.t. `NodeServicesContext`

Commit:
Merge remote-tracking branch 'origin/release/os/4.4' into vkolomeyko/CORDA-2942/node-lifecycle-events

vkolomeyko commented:
Please run smoke tests

Commit:
CORDA-2942: `NodeLifecycleObserverService` and a set of integration tests.

Public API violations are expected as well as integration tests failing.

mnesbit commented on core/src/main/kotlin/net/corda/core/node/services/NodeLifecycleObserverService.kt:
```diff
@@ -0,0 +1,31 @@
+     *      as practically possible.
+     */
+    @Throws(CordaServiceCriticalFailureException::class)
+    fun onNodeStarted() {}
```
I would very much prefer to put an observable(or registerListener style) over an event sealed class onto AppServiceHub than go this way. The problem with this is that decorating a service with an interface isn't extensible. Every time we add a new event e.g. a CordappLoadEvent we break existing apps that don't implement the new methods and probably don't care, or else we have to add more and more interfaces..

vkolomeyko replied:
Agree.
@mnesbit and I discussed offline and I will re-work.

vkolomeyko replied:
Changes pushed.

Commit:
CORDA-2942: Re-work to introduce `ServiceLifecycleObserver`

Commit:
CORDA-2942: Explicitly mention a type of exception that may be thrown for some events.

Commit:
CORDA-2942: Register `ServiceLifecycleObserver` through `AppServiceHub`

Commit:
CORDA-2942: Fix integration test + KDocs update

Commit:
CORDA-2942: Detekt and `api-current` update

Commit:
CORDA-2942: Improvement to `CordaServiceLifecycleFatalTests`

... or else it has side effects on other tests.

Commit:
CORDA-2942: Add an integration test for new API use in Java

Driver test is written in Kotlin, but services definition is written in Java.

Also KDocs improvements.

Commit:
CORDA-2942: Documentation and release notes update

mnesbit commented on .ci/api-current.txt:
```diff
@@ -3367,6 +3367,9 @@ public static final class net.corda.core.messaging.StateMachineUpdate$Removed ex
 public interface net.corda.core.node.AppServiceHub extends net.corda.core.node.ServiceHub
+  @NotNull
+  public abstract net.corda.core.node.services.vault.CordaTransactionSupport getDatabase()
+  public abstract void register(net.corda.core.node.services.ServiceLifecycleObserver, boolean)
```
I think we are meant to add the updated API in 4.5, but we should check.

vkolomeyko replied:
I can take them out if necessary. I thought it would make sense to list them since change log now mentions them both.
@rick-r3 what's your take if we should add those new methods to public API?

vkolomeyko replied:
Undone `api-current.txt` changes

mnesbit commented on core/src/main/kotlin/net/corda/core/node/services/ServiceLifecycleObserver.kt:
```diff
@@ -0,0 +1,41 @@
+     * sense for Corda node to continue its operation. The lifecycle events dispatcher will endeavor to terminate node's JVM as soon
+     * as practically possible.
+     */
+    NODE_STARTED,
```
Should this be CORDAPP_STARTED and CORDAPP_STOPPED, since we are really about their lifecycle even if currently it aligns.

vkolomeyko replied:
Sure - will change.

mnesbit commented on docs/source/api-service-classes.rst:
```diff
@@ -32,7 +32,14 @@ Below is an empty implementation of a Service class:
+                // Custom code ran at service creation
+
+                // Optional: Express interest in receiving lifecycle events
+                services.register(object : ServiceLifecycleObserver {
```
Is their a flavour of Kotlin syntax that could push it onto a member function?

vkolomeyko replied:
Yes, will change.

mnesbit commented on docs/source/api-service-classes.rst:
```diff
@@ -32,7 +32,14 @@ Below is an empty implementation of a Service class:
+                // Optional: Express interest in receiving lifecycle events
+                services.register(object : ServiceLifecycleObserver {
+                    override fun onServiceLifecycleEvent(event: ServiceLifecycleEvent) {
+                        // Lifecycle event handling code
```
We should show this doing a vault query and then starting a flow.

vkolomeyko replied:
Sure

mnesbit commented on core/src/main/kotlin/net/corda/core/node/services/ServiceLifecycleObserver.kt:
```diff
@@ -0,0 +1,41 @@
+    /**
+     * Notification to inform that Corda Node is shutting down. In response to this event [CordaService] may perform clean-up of some critical
+     * resources.
+     */
```
Need to note that this is best effort and some node failures may not call it.

vkolomeyko replied:
Sure

mnesbit commented on docs/source/api-service-classes.rst:
```diff
@@ -32,7 +32,14 @@ Below is an empty implementation of a Service class:
         class MyCordaService(private val serviceHub: AppServiceHub) : SingletonSerializeAsToken() {
```
Emphasize importance of SingletonSerializeAsToken

vkolomeyko replied:
`Creating Service` section already has: `* Extend SingletonSerializeAsToken`

mnesbit commented on docs/source/api-service-classes.rst:
```diff
@@ -43,11 +50,19 @@ Below is an empty implementation of a Service class:
+                // Optional: Express interest in receiving lifecycle events
+                serviceHub.register(new ServiceLifecycleObserver() {
+                    @Override
+                    public void onServiceLifecycleEvent(@NotNull ServiceLifecycleEvent event) {
```
as above

vkolomeyko replied:
Will change

mnesbit commented on core/src/main/kotlin/net/corda/core/node/AppServiceHub.kt:
```diff
@@ -38,4 +39,13 @@ interface AppServiceHub : ServiceHub {
+     * @param priority if set to `true` the [observer] will be added to a priority queue such that it will be notified ahead of non-prioritised
+     *      observers.
+     */
+    fun register(observer: ServiceLifecycleObserver, priority: Boolean = false)
```
perhaps a priority enum rather than a boolean, though two levels sounds fine to me?

mnesbit commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleEventsDistributor.kt:
```diff
@@ -0,0 +1,109 @@
+    /**
+     * Distributes event to all the observers previously added
+     */
+    fun distributeEvent(event: NodeLifecycleEvent) {
```
We have done this thread safe, which won't be exercised right now. So it would be wise to add a multi-threaded thrasher just to sanity check it all.

vkolomeyko replied:
Sure - will do

Commit:
CORDA-2942: First set of changes following review by @mnesbit

Commit:
CORDA-2942: Second set of changes following review by @mnesbit

mnesbit commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleEventsDistributor.kt:
```diff
@@ -0,0 +1,109 @@
+
+        if(updateFailed.isNotEmpty()) {
+            log.error("Failed to distribute event $event, failure outcome: $updateFailed")
+            if(event.javaClass in crtiticalEventsClasses) {
```
Typo

vkolomeyko replied:
Sure - thanks

mnesbit commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleEventsDistributor.kt:
```diff
@@ -0,0 +1,109 @@
+        readWriteLock.writeLock().executeLocked {
+            prioritizedObservers.addAll(observers)
+            // In-place sorting
+            prioritizedObservers.sortBy { SortingKey(it.priority, it.javaClass) }
```
Use a TreeMap?

vkolomeyko replied:
I was thinking about it, but did not want to trigger repetitive sorting when multiple items are added.

Commit:
CORDA-2942: Added multi-threaded test

Commit:
CORDA-2942: Fixes

Commit:
CORDA-2942: Undo changes to `api-current.txt`

Commit:
CORDA-2942: Bare mimimum change to `api-current.txt` for CI gate to pass.

vkolomeyko commented on .ci/api-current.txt:
```diff
@@ -3754,6 +3754,9 @@ public static final class net.corda.core.node.services.PartyInfo$SingleNode exte
   @NotNull
   public String toString()
 ##
+public interface net.corda.core.node.services.ServiceLifecycleObserver
```
Note: This change is necessary as part of this PR, or else API stability CI gate will fail.

rick-r3 commented on common/configuration-parsing/src/main/kotlin/net/corda/common/configuration/parsing/internal/Utils.kt:
```diff
@@ -64,4 +64,13 @@ internal fun ConfigValue.serialize(options: ConfigRenderOptions = ConfigRenderOp
\ No newline at end of file
+internal fun <TYPE> valid(target: TYPE) = Validated.valid<TYPE, Configuration.Validation.Error>(target)
+
+data class ConfigurationWithOptions(private val config: Config, private val options: Configuration.Validation.Options) {
```
It's worth a little bit of explanation of what these classes / interfaces are for.

vkolomeyko replied:
Sure - will do.

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/AppServiceHub.kt:
```diff
@@ -38,4 +41,22 @@ interface AppServiceHub : ServiceHub {
+     * @param priority if set to `true` the [observer] will be added to a priority queue such that it will be notified ahead of non-prioritised
+     *      observers.
+     */
+    fun register(observer: ServiceLifecycleObserver, priority: ServiceLifecycleObserverPriority = ServiceLifecycleObserverPriority.MEDIUM)
```
I would re-order the parameters here.  It bugs me that this is observer followed by priority but the convenience method is priority followed by observer.  Having the lambda on the end makes sense for kotlin, so I think this one has to change.

vkolomeyko replied:
I agree - let's keep it consistent.

rick-r3 commented on docs/source/api-service-classes.rst:
```diff
@@ -32,7 +32,16 @@ Below is an empty implementation of a Service class:
+            }
+
+            private fun processEvent(event: ServiceLifecycleEvent) {
+                // Lifecycle event handling code including full use of serviceHub
```
The example should probably not just treat all events the same, as that is probably a bad example.   It should do a switch on the event type, no ?  

Should there always be an `else` in the switch (see question above about possibility of additional event enum entries) ?

Will `ServiceHub` always be available ?  Probably okay to say this if we never expand the event types, but maybe it's a property of particular events (if there are going to be more).

vkolomeyko replied:
For the limited number of events we expose to `CordaService` (i.e. `CORDAPP_STARTED` and `CORDAPP_STOPPED`) `ServiceHub` will always be available.

I do agree though it would make to expand this example and show how to correctly process various event types.
Agree it will make sense to add `else` into the switch.

rick-r3 commented on docs/source/api-service-classes.rst:
```diff
@@ -43,11 +52,20 @@ Below is an empty implementation of a Service class:
+            }
+
+            private void processEvent(ServiceLifecycleEvent event) {
+                // Lifecycle event handling code including full use of serviceHub
```
Same comment as above.

vkolomeyko replied:
Will change.

rick-r3 commented on docs/source/api-service-classes.rst:
```diff
@@ -59,9 +77,9 @@ from ``AppServiceHub`` is explained further in :ref:`Starting Flows from a Servi
-This is useful when a service needs to establish a connection to an external database or setup observables via ``ServiceHub.trackBy`` during
-its startup. These can then persist during the service's lifetime.
+Also the ``AppServiceHub`` provides ability for ``CordaService`` to subscribe for lifecycle events of the node, such that it will get notified
+about node finished initialisation and when the node is shutting down such that ``CordaService`` will be able to perform clean-up of some
```
`node finishing`

vkolomeyko replied:
Will change

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/services/ServiceLifecycleObserver.kt:
```diff
@@ -0,0 +1,52 @@
+    HIGH, MEDIUM, LOW
+}
+
+enum class ServiceLifecycleEvent {
```
Are we intending this set of events will expand over time?

vkolomeyko replied:
Potentially "yes", but since it is a public API we have to be careful of what we are exposing to external users such that it is not changing too often.

vkolomeyko replied:
OK, `ServiceLifecycleObserverPriority` been deleted now and will probably expand enum `ServiceLifecycleEvent` going forward.

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleEventsDistributor.kt:
```diff
@@ -0,0 +1,109 @@
+        val snapshot = readWriteLock.readLock().executeLocked { LinkedList(prioritizedObservers) }
+        val orderedSnapshot = if (event.reversedPriority) snapshot.reversed() else snapshot
+
+        val updateResult = orderedSnapshot.map { it.update(event) }
```
I'm not sure it makes sense to carry on in the event of a failure.  We should fail fast, rather than analysing at the end.  Who knows what harm can be caused by continuing partially initialised.

vkolomeyko replied:
OK, I will change this logic.

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleEventsDistributor.kt:
```diff
@@ -0,0 +1,109 @@
+        }
+    }
+
+    private fun <T> Lock.executeLocked(block: () -> T) : T {
```
Worth a comment on why we have our implementation here.  So someone doesn't refactor back to using the stdlib one.

vkolomeyko replied:
Sure

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleObserver.kt:
```diff
@@ -0,0 +1,57 @@
+    fun update(nodeLifecycleEvent: NodeLifecycleEvent) : Try<String> = Try.on { "${javaClass.simpleName} ignored $nodeLifecycleEvent" }
+
+    /**
+     * It is possibly to optionally override observer priority.
```
`possibly` -> `possible`

vkolomeyko replied:
Will change.

rick-r3 commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleObserver.kt:
```diff
@@ -0,0 +1,57 @@
+ * A set of events to flag the important milestones in the lifecycle of the node.
+ * @param reversedPriority flags whether it would make sense to notify observers in the reversed order.
+ */
+sealed class NodeLifecycleEvent(val reversedPriority: Boolean = false) {
```
Rather than trying to manage dependencies via priorities (e.g. the RPC vs. SERVICE priorities above), we should have more life cycle stages / events to represent any important milestones during node start (or shutdown).  Doing it via priorities will be harder to manage.

Do we need to use the different priorities now or later? If later, let's just remove the different priorities above and just have high/default/low, to avoid implying we're going to manage dependencies using priorities.  If now, let's figure out what additional events we need.

(The original point of the priorities was to just allow any adjustment of ordering should it be absolutely necessary for extensions.  If we learn about the need for more events to help avoid that, let's introduce the events / life cycle stages).

vkolomeyko replied:
I would say "later".

As discussed with @mnesbit the idea was to define few useful constants of `SERVICE` and `RPC` families to ensure that services are always receiving lifecycle events ahead of RPC implementations.
Moreover, in our public API for the lifecycle for the services we only declare enum `ServiceLifecycleObserverPriority`, so the end users are not exposed to all those `int`s and will not be able to set any unreasonable value.

vkolomeyko replied:
As discussed further with @mnesbit this morning I have deleted `ServiceLifecycleObserverPriority` and switched to `Int`s I have also introduced two new event types: `CorDappStarted` and `CorDappStopped`.

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/services/ServiceLifecycleObserver.kt:
```diff
@@ -0,0 +1,52 @@
+ * Controls lifecycle event distribution priority.
+ * Observers of any priority will eventually will be notified and this enum helps to control the notification order.
+ */
+enum class ServiceLifecycleObserverPriority {
```
Let's just make the priority an `int`.

vkolomeyko replied:
As discussed above, this is part of public API. @mnesbit and I discussed and we thought it would make sense to constrain it so that end-users will not be able to pass an unreasonable value.

**Note:** Internal lifecycle observers `NodeLifecycleObserver` have their `priority` as `int` - but this is for internal use and people tinkering with those will know more about internals of Corda Node.

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/AppServiceHubImpl.kt:
```diff
@@ -0,0 +1,108 @@
+        private class NodeLifecycleServiceObserverAdaptor(private val observer: ServiceLifecycleObserver, priorityEnum: ServiceLifecycleObserverPriority) : NodeLifecycleObserver {
+            override val priority: Int = when(priorityEnum) {
+                ServiceLifecycleObserverPriority.LOW -> SERVICE_PRIORITY_LOW
+                ServiceLifecycleObserverPriority.MEDIUM -> SERVICE_PRIORITY_NORMAL
```
I don't think we should have this complex mapping.  Just use `int` and pass through.  Whatever we do, we'll end up wanting something to be lower priority than something that is `LOW`.  Using an `int` means you can always go lower (unless someone unwisely used `MIN_VALUE`).

vkolomeyko replied:
Please see above my comment about `enum` vs. `int`.

Commit:
CORDA-2942: Address review feedback from @rick-r3

Commit:
Merge remote-tracking branch 'origin/release/os/4.4' into vkolomeyko/CORDA-2942/node-lifecycle-events

# Conflicts:
#	docs/source/changelog.rst

Commit:
CORDA-2942: Detekt update

Commit:
CORDA-2942: Delete `ServiceLifecycleObserverPriority` and replace it with `Int` after discussion with @mnesbit

Commit:
CORDA-2942: Introduce more `NodeLifecycleEvent` and switch services to listen for those events

mnesbit commented:
I am fine with the API aspects of this. We need an integration test to confirm that the basic start event can read the vault and start a flow successfully. The shutdown should be able to write an item into a DB table. We may need to tests to verify we don't get any double firing of the events at either CorDapp, or internal levels just in case that regresses at a future date.

rick-r3 commented on node/src/integration-test/java/net/corda/node/services/JavaCordaServiceLifecycle.java:
```diff
@@ -0,0 +1,69 @@
+            switch (event) {
+                case CORDAPP_STARTED:
+                    eventsCaptured.add(event);
+                    // serviceHub.startFlow()
```
Commented out ?

vkolomeyko replied:
I will tidy-up.
I was using this code to copy paste into docs just to check there are no syntax errors.

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -518,7 +526,12 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
+            // Wait for SMM to fully start
+            smmStartedFuture.get()
+            nodeLifecycleEventsDistributor.distributeEvent(NodeLifecycleEvent.AfterNodeStart(nodeServicesContext))
+            nodeLifecycleEventsDistributor.distributeEvent(NodeLifecycleEvent.CorDappStarted(nodeServicesContext))
```
I'm right in thinking (from the tests) that this fires _after_ flows start?
What would a CordaService listen for if it wanted to do something in order to be ready for flows that needed AppServiceHub?  I think we need two events.  `ServiceHubReady` or something?

rick-r3 replied:
Actually, maybe this doesn't matter.  They can use the _rest_ of `AppServiceHub` at construction time.  So extra event not needed.

vkolomeyko replied:
Agree - `@CordaService` has rich `AppServiceHub` facility at construction time.

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/AppServiceHub.kt:
```diff
@@ -38,4 +46,24 @@ interface AppServiceHub : ServiceHub {
+    /**
+     * Allows to register [ServiceLifecycleObserver] such that it will start receiving [net.corda.core.node.services.ServiceLifecycleEvent]s
+     *
+     * @param priority controls to which queue [observer] will be added. Lower values correspond to higher priorities.
```
Should we maybe reverse the sort so that higher `priority` is higher priority.

vkolomeyko replied:
I do not mind, but here is another idea: How about changing word `priority` to `order` - in this case, I think, it would be OK to have lower value to be at the "front of the queue".

rick-r3 replied:
Not sure I like that.   It was a good idea to try and come up with another word, but `priority` seems right to me.

rick-r3 replied:
I'm not obsessed with fixing this if it's a big upheaval, but I thought it would be a case of changing the three constants and inverting the sort.

vkolomeyko replied:
I will do it. It is indeed a small change to make. Let's make it right before we release it into the wild :-)

Commit:
First stub on integration test
Unfinished - hang on issue and pay

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/AppServiceHubImpl.kt:
```diff
@@ -0,0 +1,98 @@
+        require(logicType.isAnnotationPresent(StartableByService::class.java)) { "${logicType.name} was not designed for starting by a CordaService" }
+        // TODO check service permissions
+        // TODO switch from myInfo.legalIdentities[0].name to current node's identity as soon as available
+        val context = InvocationContext.service(serviceInstance.javaClass.name, myInfo.legalIdentities[0].name)
```
We could check in here that the correct stage of the lifecycle has been reached and provide a helpful error (or just log a warning)!  However, what exactly happens today if you call `startFlow` ?  I've forgotten what the behaviour is.  It _might_ be that it works, unless you wait for the returned `Future`.

vkolomeyko replied:
What is currently happening is largely depends on luck and whether people apply workaround like doing "sleep" or spawning background thread from constructor.

I agree that issuing a warning could be helpful, however currently for a given `@CordaService` there is no uniform API to ask "Which lifecycle events you have received so far?"
Unless we prepared to include this functionality in public API I am not sure I see how this can be done.

rick-r3 replied:
Well, my point was `AppServiceHubImpl` could subscribe to events and then set a flag once it sees the state machine started event.  If that flag is not set when someone calls this method, then issue the warning.

vkolomeyko replied:
I see your point. Yes, indeed, `AppServiceHubImpl` may as well become as subscriber of lifecycle events and have a guard against dispatching flows too soon.

rick-r3 commented on core/src/main/kotlin/net/corda/core/node/services/ServiceLifecycleObserver.kt:
```diff
@@ -0,0 +1,44 @@
+enum class ServiceLifecycleEvent {
+    /**
+     * This event is dispatched when CorDapp is fully started such that [net.corda.core.node.AppServiceHub] available
+     * for [CordaService] to be use.
```
`This event is dispatched when the state machine is fully started such that [net.corda.core.node.AppServiceHub] is able to support methods that start flows.`  All other methods work anyway, but see my comment later about actually understanding the behaviour of the start flow methods.  Can they be invoked?  Or is it just the waiting for the `Future` that we cannot do ?

mnesbit replied:
Historically the corda service constructor doesn't work for all sorts of bits and pieces of servicehub

vkolomeyko replied:
I agree with @mnesbit - `AppServiceHub` is unsafe to use at `CordaService` construction time.

Commit:
CORDA-2942: Few more changes after input from @rick-r3

Commit:
CORDA-2942: Switch to use out-of-process nodes for the inetgration test

Currently Alice and Notary stuck waiting to hear from each other.

Commit:
CORDA-2942: Extra log lines during event distribution

Commit:
CORDA-2942: Asynchronously distribute lifecycle events

Commit:
CORDA-2942: Await for complete P2P client start-up

Next step: Add vault query to integration test

Commit:
CORDA-2942: Asynchronously distribute lifecycle events

Next step: Improve integration test

Commit:
CORDA-2942: Fix test broken by recent changes and improve logging

rick-r3 reviewed (commented):
Apparently I had comments pending from Friday

Commit:
CORDA-2942: Improvement of the test to be able to monitor actions performed by @CordaService in the remote process

Commit:
CORDA-2942: Add node re-start step to the integration test

Commit:
CORDA-2942: Remove `CORDAPP_STOPPED` event for now

Commit:
CORDA-2942: s/CORDAPP_STARTED/STATE_MACHINE_STARTED/

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt:
```diff
@@ -505,21 +515,33 @@ abstract class AbstractNode<S>(val configuration: NodeConfiguration,
+            // NB: Dispatch lifecycle events outside of transaction to ensure attachments and the like persisted into the DB
+            log.debug("Distributing events")
+            nodeLifecycleEventsDistributor.distributeEvent(NodeLifecycleEvent.AfterNodeStart(nodeServicesContext))
+            nodeLifecycleEventsDistributor.distributeEvent(NodeLifecycleEvent.CorDappStarted(nodeServicesContext))
```
I think I would be in favour of *not* calling this `CorDappStarted` but something like `StateMachineStarted`.  I think that is clearer.

vkolomeyko replied:
Sure - will change this.

Commit:
CORDA-2942: Inverse the meaning of `priority` as requested by @rick-r3

Commit:
CORDA-2942: Register `AppServiceHubImpl` for lifecycle events and put a warning when SMM is not ready.

vkolomeyko commented:
Please run smoke tests

rick-r3 commented on node/src/main/kotlin/net/corda/node/internal/AppServiceHubImpl.kt:
```diff
@@ -0,0 +1,124 @@
+        // TODO switch from myInfo.legalIdentities[0].name to current node's identity as soon as available
+        if(!flowsAllowed) {
+            logger.warn("Flow $flow started too early in the node's lifecycle, SMM may not be ready yet. " +
+                    "Please consider registering your service to node's lifecycle event: `STATE_MACHINE_STARTED`")
```
I think this should be service lifecycle ?  Fix in a follow up.

vkolomeyko replied:
`NodeLifecycleEvent.StateMachineStarted` is an internal event whereas `STATE_MACHINE_STARTED` is the enum value we expose to the end-users.
So if warning is produced that meant that someone is starting a flow too soon, e.g. from constructor of `@CordaService`.
Hence comment is correct, as it encourages people to stop coding logic in the constructor of `@CordaService` but instead switch to handle `STATE_MACHINE_STARTED` events.

rick-r3 reviewed (approved):
Okay with me.  See one comment for a follow up.
